### PR TITLE
dockerify: override settings from environment

### PR DESCRIPTION
**Pull request by khoutz182**
The main use case i have in mind for this is for docker deployments, where we can have a static docker image, most of the defaults defined in `environment.json` but overrides are possible in every case through environment variables.

### CODE REVIEW

File: src/test/java/com/elmsoftware/env/EnvironmentSettingsModuleTest.java [hunk 1]
```diff
@@ -118,17 +97,23 @@ public void should_use_settings_provider_if_provided() throws Exception {
...
+		// setup test
+		final String dockerEnvironmentVariable = "I'm configurable for docker!";
+		final EnvironmentSettingsModule module = new EnvironmentSettingsModule(new NoOpSettingProvider(), util);
 
+		System.setProperty("environment.json", "guice-test.json");
+		when(binder.bind(any(Key.class))).thenReturn(builder);
+		when(binder.skipSources(Names.class)).thenReturn(binder);
+
+		// run test
+		withEnvironmentVariable("SOME_KEY", dockerEnvironmentVariable)
```
Comment (khoutz182) [inline src/test/java/com/elmsoftware/env/EnvironmentSettingsModuleTest.java:127]: im usually not a fan of bringing in a new dependency for one test, but in this case it seemed easiest.

File: src/main/java/com/elmsoftware/env/EnvironmentSettingsModule.java [hunk 2]
```diff
@@ -55,4 +57,18 @@ protected void configure(final Binder binder, final Map<String, String> properti
...
 
+	protected void overrideEnvironmentSettings(final EnvironmentSettings environmentSettings, final String environment) {
+		final Map<String, String> settings = environmentSettings.getEnvironmentSettings().get(environment);
+
+		settings.keySet()
+			.forEach(key -> {
+				final String envKey = key.replaceAll("[^A-Za-z0-9]", "_").toUpperCase(Locale.US);
+				final String envValue = System.getenv(envKey);
+				if (envValue != null) {
+					log.debug("Overriding " + key + " with environment variable");
```
Comment (khoutz182) [inline src/main/java/com/elmsoftware/env/EnvironmentSettingsModule.java]: i originally wrapped this with a debug enabled check:
```suggestion
					if (log.isDebugEnabled()) {
						log.debug("Overriding " + key + " with environment variable");
					}
```
but getting the code coverage for that was going to be annoying. generating that string isn't too expensive at least

Comment (lmeadors) [reply src/main/java/com/elmsoftware/env/EnvironmentSettingsModule.java]: Use the `{}` format for logging, it's pretty cheap and clean; all it has to do is push the reference onto the stack, then it decides of it needs to build the string or not later.

File: src/main/java/com/elmsoftware/env/EnvironmentSettingsModule.java [hunk 3]
```diff
@@ -44,15 +56,31 @@ public void configure(final Binder binder) {
...
-		final Map<String, String> properties = settings.merge(environment, settingProvider);
+		final Properties properties = new Properties();
+		properties.putAll(settings.merge(environment, settingProvider));
+		settingPostProcessor.process(properties);
 
 		Names.bindProperties(binder, properties);
 		configure(binder, properties);
+	}
 
+	protected void configure(final Binder binder, final Properties properties) {
```
Comment (khoutz182) [inline src/main/java/com/elmsoftware/env/EnvironmentSettingsModule.java:81]: the move from a map to properties makes sense to me. I deprecated the map implementation so we can feel okay with removing it on a later release, but left a default implementation on the properties version so there's no api change with this.

File: src/test/java/com/elmsoftware/env/EnvironmentSettingsModuleTest.java [hunk 4]
```diff
@@ -118,17 +108,55 @@ public void should_use_settings_provider_if_provided() throws Exception {
...
+		verify(builder).toInstance(eq(dockerEnvironmentVariable));
 	}
 
-	private Key<String> findKey(final List<Key<String>> allValues, final String name) {
-		for (final Key<String> key : allValues) {
-			final Named annotation = (Named) key.getAnnotation();
-			if (annotation.value().equals(name)) {
-				return key;
+	@Test
+	public void should_use_legacy_configure_on_only_string_properties() {
```
Comment (khoutz182) [inline src/test/java/com/elmsoftware/env/EnvironmentSettingsModuleTest.java:135]: this is a weird test, and i'm not proud of it. it's goal is mostly just to make sure property values that aren't strings don't get passed in to the legacy `configure(Binder, Map<String, String>)` implementation.

**Pull request closed without merging**